SYS_getrandom undeclared compiling nim 1.6.0 stdlib on linux kernel < 3.17 (including RHEL7)

On kernel versions < 3.17, compilation of nim fails at the point of compiling std/sysrand with a message similar to the below:
```
/opt/src/nim/testament_r/stdlib_sysrand.nim.c: In function ‘urandomInternalImpl__stdZsysrand_17’:
/opt/src/nim/testament_r/stdlib_sysrand.nim.c:61:35: error: ‘SYS_getrandom’ undeclared (first use in this function); did you mean ‘SYS_getuid’?
    readBytesX60gensym0_ = syscall(SYS_getrandom, ((void*) ((&dest[result]))), ((int) ((NI)(TM__sT0fWggTrXhnxQ79bl9a8nHg_2))), ((unsigned int) 0));
```

The reason for this becomes obvious pretty quickly if you're familiar with getrandom in the kernel, since it doesn't exist until version 3.17. PR is incoming with a _very_ simple fix, resolving the issue by avoiding use of getrandom when passed an associated compile flag (more on this below).

Detailed Description:

As of now, Linux distros that use older kernels (such as 3.10 in RHEL7 for example), cannot compile and use nim versions >= 1.6.0. This is because on linux kernel versions < 3.17, most notably including any/all RHEL (and CentOS) 7.x based systems, the syscall for getrandom doesn't exist. The latest stable release (1.6.0) includes sysrand in stdlib, and the missing the syscall to getrandom in the 3.10 kernel appears to be the _sole_ issue preventing functional nim compilation and usage by users of this OS/distro (presuming that GCC 8 is used via `yum install devtoolset-8` and associated initialization is performed via `scl enable devtoolset-8 bash`).

Fallback logic exists in std/sysrand which uses /dev/urandom as a PRNG, however this fallback logic is only performed if nim does not recognize it's being compiled on Linux (the logic is done for any unknown POSIX OS). If, as is the case here, nim does recognize it's being compiled on Linux, it simply uses the template for linux and fails on a missing enum for the syscall of getrandom. The fallback logic, however, provides a ready made path to resolve the issue. It should (IMHO, see below) be put behind a compile flag. This also has the added benefit of causing effectively no potential impact to already functional code/build systems.

Of course, getrandom is used because it's a cryptographically secure PRNG (Pseudo Random Number Generator) along with /dev/urandom, though it's preferred for several reasons over using /dev/urandom (for examples see [here](https://lwn.net/Articles/606141/) at the introduction of getrandom for some), however usage of /dev/urandom remains the fallback for a POSIX OS which can't be relied upon to include getrandom. Determining the presence of getrandom at runtime may be possible (I'll admit I started down that road, had several unexpected issues with runtime detection of var declaration via importc, then thought better of the entire approach). In this case, I'm not convinced it would be preferable to detect the presence of the getrandom syscall at runtime given it's intended for use as a cryptographically secure PRNG. I for one can't say that a runtime determination wouldn't expose an exploit of some kind, given I'd be doing so using likely a somewhat hamfisted approach, whereas a compile flag instead explicitly (on linux) either uses getrandom or uses the fallback of /dev/urandom. A compile flag also guarantees consistency in PRNG logic for a given binary, something that is again likely preferable when considering it's a PRNG used for cryptographic purposes (i.e. crashing on getrandom access vs falling back to /dev/urandom).

If we wanted to make this dynamic despite the above, I'll need a bit of guidance on how best to do so. When declaring a var via importc as with `var SYS_getrandom {.importc: "SYS_getrandom", header: "<sys/syscall.h>".}: clong`, subsequent calls to `declared(SYS_getrandom)` return true, however actually attempting to use the variable (i.e. `if SYS_getrandom == 0:`) throws an undeclared identifier error. As previously stated, I'd suggest it's actually preferable to have this behind a compile flag rather than a runtime or automatic determination at compile time, but I can reach out to discuss how to resolve this if my opinion differs from that of Araq or others.

--- a/changelogs/changelog.md
+++ b/changelogs/changelog.md
@@ -1,0 +1,31 @@
+# v1.xx.x - yyyy-mm-dd
+
+## Changes affecting backward compatibility
+
+## Standard library additions and changes
+
+### New compile flag (`-d:nimNoGetRandom`) when building `std/sysrand` to remove dependency on linux `getrandom` syscall
+
+This compile flag only affects linux builds and is necessary if either compiling on a linux kernel version < 3.17, or if code built will be executing on kernel < 3.17.
+
+On linux kernels < 3.17 (such as kernel 3.10 in RHEL7 and CentOS7), the `getrandom` syscall was not yet introduced. Without this, the `std/sysrand` module will not build properly, and if code is built on a kernel >= 3.17 without the flag, any usage of the `std/sysrand` module will fail to execute on a kernel < 3.17 (since it attempts to perform a syscall to `getrandom`, which isn't present in the current kernel). A compile flag has been added to force the `std/sysrand` module to use /dev/urandom (available since linux kernel 1.3.30), rather than the `getrandom` syscall. This allows for use of a cryptographically secure PRNG, regardless of kernel support for the `getrandom` syscall.
+
+When building for RHEL7/CentOS7 for example, the entire build process for nim from a source package would then be:
+```sh
+$ yum install devtoolset-8 # Install GCC version 8 vs the standard 4.8.5 on RHEL7/CentOS7. Alternatively use -d:nimEmulateOverflowChecks. See issue #13692 for details
+$ scl enable devtoolset-8 bash # Run bash shell with default toolchain of gcc 8
+$ sh build.sh  # per unix install instructions
+$ bin/nim c koch  # per unix install instructions
+$ ./koch boot -d:release  # per unix install instructions
+$ ./koch tools -d:nimNoGetRandom  # pass the nimNoGetRandom flag to compile std/sysrand without support for getrandom syscall
+```
+
+This is necessary to pass when building nim on kernel versions < 3.17 in particular to avoid an error of "SYS_getrandom undeclared" during the build process for stdlib (sysrand in particular).
+
+## Language changes
+
+
+## Compiler changes
+
+
+## Tool changes

--- a/lib/std/sysrand.nim
+++ b/lib/std/sysrand.nim
@@ -159,7 +159,7 @@ elif defined(windows):
 
     result = randomBytes(addr dest[0], size)
 
-elif defined(linux):
+elif defined(linux) and not defined(nimNoGetRandom):
   # TODO using let, pending bootstrap >= 1.4.0
   var SYS_getrandom {.importc: "SYS_getrandom", header: "<sys/syscall.h>".}: clong
   const syscallHeader = """#include <unistd.h>

--- a/lib/std/sysrand.nim
+++ b/lib/std/sysrand.nim
@@ -38,6 +38,11 @@
 ## .. _randomFillSync: https://nodejs.org/api/crypto.html#crypto_crypto_randomfillsync_buffer_offset_size
 ## .. _/dev/urandom: https://en.wikipedia.org/wiki//dev/random
 ##
+## On a Linux target, a call to the `getrandom` syscall can be avoided (e.g.
+## for targets running kernel version < 3.17) by passing a compile flag of
+## `-d:nimNoGetRandom`. If this flag is passed, sysrand will use `/dev/urandom`
+## as with any other POSIX compliant OS.
+##
 
 runnableExamples:
   doAssert urandom(0).len == 0

